[LFC][Integration] Add ability to specify height constraint for layout

sammygill opened the pull request:
#### 11347fc8b9726376fa9de1dfa8cb34c77ecb1ea3
<pre>
[LFC][Integration] Add ability to specify height constraint for layout
<a href="https://bugs.webkit.org/show_bug.cgi?id=299982">https://bugs.webkit.org/show_bug.cgi?id=299982</a>
<a href="https://rdar.apple.com/161762080">rdar://161762080</a>

Reviewed by Alan Baradlay.

Currently, we have the ability to perform layout on a box with a width
constraint that gets set on the renderer via integration. This
constraint can then get used by the formatting context generated by the
box for sizing its contents. In many cases, grid layout is able to
provide a definite size for the width and height of its items, which
should in turn act as constraints for the grid items’ formatting context.
For example, a stretching grid item will fill the available space from
its grid area. Likewise, percentages will be able to resolve against
sizes in both dimensions.

In preparation for being able to perform layout on grid items as one of
the final steps of grid layout, we can expand this API to optionally
take in a height constraint. Callers can then pass in any definite
constraints, which should be determined from the formatting context
driving layout.

(WebCore::Layout::LayoutState::layoutWithFormattingContextForBox const):
(WebCore::LayoutIntegration::layoutWithFormattingContextForBox):
(WebCore::Layout::IntegrationUtils::layoutWithFormattingContextForBox const):
(WebCore::Layout::IntegrationUtils::layoutWithFormattingContextForBox):
All basically just plumbing for the new constraint.

Canonical link: <a href="https://commits.webkit.org/300908@main">https://commits.webkit.org/300908@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/521aa8675ac6be744a80848044e57ee92e083cff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43952 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34676 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131102 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/04bc3e22-5104-479d-9169-6d19e051f264) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126143 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52554 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94523 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/63e06733-239b-41c4-8628-f36a5b409235) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111153 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75112 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9358f124-fe20-4af8-854e-28d7aab1e04b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34566 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29312 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74583 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105365 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29534 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133773 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39020 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102999 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51570 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107369 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102799 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26160 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48163 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26417 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51041 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56823 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50480 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53836 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52155 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/LayoutState.h:
```diff
@@ -114,7 +114,7 @@ class LayoutState : public CanMakeWeakPtr<LayoutState>, public CanMakeThreadSafe
     const ElementBox& root() const { return m_rootContainer; }
 
-    void layoutWithFormattingContextForBox(const ElementBox&, std::optional<LayoutUnit> widthConstraint) const;
+    void layoutWithFormattingContextForBox(const ElementBox&, std::optional<LayoutUnit> widthConstraint, std::optional<LayoutUnit> heightConstraint) const;
```
This is slightly misleading as it would only work on a few selected type of boxes, so at least I would add a comment explaining it.

sammygill committed:
[LFC][Integration] Add ability to specify height constraint for layout
https://bugs.webkit.org/show_bug.cgi?id=299982
rdar://161762080

Reviewed by Alan Baradlay.

Currently, we have the ability to perform layout on a box with a width
constraint that gets set on the renderer via integration. This
constraint can then get used by the formatting context generated by the
box for sizing its contents. In many cases, grid layout is able to
provide a definite size for the width and height of its items, which
should in turn act as constraints for the grid items’ formatting context.
For example, a stretching grid item will fill the available space from
its grid area. Likewise, percentages will be able to resolve against
sizes in both dimensions.

In preparation for being able to perform layout on grid items as one of
the final steps of grid layout, we can expand this API to optionally
take in a height constraint. Callers can then pass in any definite
constraints, which should be determined from the formatting context
driving layout.

(WebCore::Layout::LayoutState::layoutWithFormattingContextForBox const):
(WebCore::LayoutIntegration::layoutWithFormattingContextForBox):
(WebCore::Layout::IntegrationUtils::layoutWithFormattingContextForBox const):
(WebCore::Layout::IntegrationUtils::layoutWithFormattingContextForBox):
All basically just plumbing for the new constraint.

Canonical link: https://commits.webkit.org/300908@main
